### PR TITLE
Support for direct push to master

### PR DESCRIPTION
**Pull request by hayashikun**
Sparrow sends PRs every time, but the notifications are a little bit noisy.
It can commit and push changes to the master directly.
So I added an option, `bypass_pull_request`, to push to the master directly.

### CODE REVIEW

File: lib/sparrow/jobs/git_ops.rb [hunk 1]
```diff
@@ -48,10 +49,13 @@ def build_rewrites(target)
             source_repo: target["source_repo"],
             config_repo: target["config_repo"],
             erb_path: rewrite["erb_path"],
-            out_path: rewrite["out_path"]
+            out_path: rewrite["out_path"],
+            # if not specified, it makes PR.
+            make_pr: rewrite["make_pr"] || true
```
Comment (shouichi) [inline lib/sparrow/jobs/git_ops.rb:54]: s/make_pr/create_pull_request/g

File: lib/sparrow/jobs/git_ops/rewrite.rb [hunk 2]
```diff
@@ -77,12 +70,16 @@ def github_app_source_repo
         end
 
         def no_changes?
-          comparision = client.compare(
+          comparison = client.compare(
```
Comment (shouichi) [inline lib/sparrow/jobs/git_ops/rewrite.rb:73]: nice catch :+1:

File: lib/sparrow/jobs/git_ops/rewrite.rb [hunk 3]
```diff
@@ -39,16 +40,8 @@ def run
...
 
-          begin
-            pr = create_pull_request
-            logger.info("created a pull request", pr: pr.to_h)
-            pr
-          rescue Octokit::UnprocessableEntity => e
-            logger.info("pull request exists, skipping", error: e)
-            nil
-          end
+          push_change
```
Comment (shouichi) [inline lib/sparrow/jobs/git_ops/rewrite.rb:43]: s/push_change/create_pull_request_or_push_to_master/g

File: lib/sparrow/jobs/git_ops/rewrite.rb [hunk 4]
```diff
@@ -141,14 +138,30 @@ def create_pull_request
...
+        def push_master
+          client.update_ref(@config_repo, branch_name, commit.sha).tap do |ref|
+            logger.debug("update ref", ref: ref.to_h)
+          end
+          commit
+        end
+
+        def push_change
+          change = make_pr? ? create_pull_request : push_master
+          logger.info("created a #{make_pr? ? 'pull request' : 'commit'}", pr: change.to_h)
```
Comment (shouichi) [inline lib/sparrow/jobs/git_ops/rewrite.rb:156]: Why don't we move logging to create_pull_request/push_master?

File: lib/sparrow/jobs/git_ops.rb [hunk 5]
```diff
@@ -13,7 +13,7 @@ module Jobs
     #   1. Evaluates the given ERB file.
     #   2. Writes the result to the given output file.
     #   3. Commits the output file.
-    #   4. Creates a pull request.
+    #   4. Creates a pull request or commit to master.
```
Comment (shouichi) [inline lib/sparrow/jobs/git_ops.rb:16]: nit: s/commit/commits/

**Comment by shouichi** [comment]
@hayashikun could you describe the new feature in the readme? If a GitHub feature needs to be enabled, please describe that too. Thanks.

File: lib/sparrow/jobs/git_ops.rb (hunk 1 above)
Comment (hayashikun) [reply lib/sparrow/jobs/git_ops.rb]: The option was renamed to "create_pull_request", and the name of var/arg was set to "create_pr".

File: lib/sparrow/jobs/git_ops.rb (hunk 5 above)
Comment (hayashikun) [reply lib/sparrow/jobs/git_ops.rb]: It was written in the above line as "3. Commits the output file.", so "4. Creates a pull request or pushes to master." would be correct.

**Comment by hayashikun** [comment]
I updated README.md and sparrow.example.yml.

File: sparrow.example.yml [hunk 6]
```diff
@@ -26,3 +26,4 @@ jobs:
             - name: dev
               erb_path: overlays/dev/deploy.yaml.erb
               out_path: overlays/dev/deploy.yaml
+              create_pull_request: true
```
Comment (shouichi) [inline sparrow.example.yml:29]: github uses "bypass required pull requests", so let's use that. what about `bypass_pull_request`?

https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/managing-a-branch-protection-rule

File: README.md [hunk 7]
```diff
@@ -49,7 +49,10 @@ When a commit is pushed to the code repository's master branch,
 
 1. Cloud Build builds a container image where its tag is git SHA.
 1. The event is published to Cloud PubSub.
-1. Sparrow creates a tag rewrite pull request to the config repository.
+1. Sparrow rewrites the tag in the config repository and creates a pull request or pushes the change to the master branch.
```
Comment (shouichi) [inline README.md:52]: We also want to clarify that to enable direct pushing, we need to enable "bypass required pull requests" feature on github. Can you describe that (in a different section)?

https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/managing-a-branch-protection-rule

**Review by shouichi — COMMENTED** [review]
The code looks good :+1:

File: sparrow.example.yml (hunk 6 above)
Comment (hayashikun) [reply sparrow.example.yml:29]: sure, should I change var/arg names in Ruby code to `bypass_pull_request` from `create_pull_request`?

File: README.md (hunk 7 above)
Comment (hayashikun) [reply README.md:52]: Make sense :)
I'll add a new section about the branch protection rule.

File: sparrow.example.yml (hunk 6 above)
Comment (shouichi) [reply sparrow.example.yml:29]: Yes. Let's be consistent.

Comment (hayashikun) [reply sparrow.example.yml]: I've changed option, args, vars of `create_pull_request` to `bypass_pull_request`

File: README.md (hunk 7 above)
Comment (hayashikun) [reply README.md:52]: I've added a sentence about the branch protection rule.

File: lib/sparrow/jobs/git_ops.rb [hunk 8]
```diff
@@ -51,7 +51,7 @@ def build_rewrites(target)
             erb_path: rewrite["erb_path"],
             out_path: rewrite["out_path"],
             # if not specified, it makes PR.
-            create_pull_request: rewrite["create_pull_request"] || true
+            bypass_pull_request: rewrite["bypass_pull_request"] || true
```
Comment (shouichi) [inline lib/sparrow/jobs/git_ops.rb:54]: bypass_pull_request should be false by default

Comment (hayashikun) [reply lib/sparrow/jobs/git_ops.rb]: oh, sorry for that
I fixed it!

**Pull request merged**